sql/logic_test.go

```diff
@@ -0,0 +1,318 @@
+	case string:
+		*v = logicValue(t)
+	case time.Time:
+		*v = logicValue(t.String())
+	default:
+		return fmt.Errorf("unexpected type: %T", value)
+	}
+	return nil
+}
+
+// logicTest executes the test cases specified in a file. The file format is
+// taken from the sqllogictest tool
+// (http://www.sqlite.org/sqllogictest/doc/trunk/about.wiki) with various
+// extensions to allow specifying errors and additional options. See
+// https://github.com/gregrahn/sqllogictest/ for a github mirror of the
+// sqllogictest source.
+//
+// TODO(pmattis): We currently cannot run the tests from sqllogictest due to
+// insufficient SQL coverage (e.g. lack of subqueries and aggregation
+// functions). We should work towards fixing that.
+type logicTest struct {
+	*testing.T
+}
+
+func (t logicTest) run(path string) {
+	file, err := os.Open(path)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer file.Close()
+	base := filepath.Base(path)
+
+	// TODO(pmattis): Add a flag to make it easy to run the tests against a local
+	// MySQL or Postgres instance.
+	srv := server.StartTestServer(nil)
+	db, err := sql.Open("cockroach", "https://root@"+srv.ServingAddr()+"?certs=test_certs")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer db.Close()
+	defer srv.Stop()
+
+	s := newLineScanner(file)
+	for s.Scan() {
+		fields := strings.Fields(s.Text())
+		if len(fields) == 0 {
+			continue
+		}
+		if strings.HasPrefix(fields[0], "#") {
+			// Skip comment lines.
+			continue
+		}
+		switch fields[0] {
+		case "statement":
+			stmt := logicStatement{pos: fmt.Sprintf("%s:%d", base, s.line)}
+			// Parse "query error <regexp>"
+			if m := errorRE.FindStringSubmatch(s.Text()); len(m) == 2 {
+				stmt.expectErr = m[1]
+			}
+			var buf bytes.Buffer
+			for s.Scan() {
+				t := s.Text()
+				if t == "" {
+					break
+				}
+				fmt.Fprintln(&buf, s.Text())
+			}
+			stmt.sql = strings.TrimSpace(buf.String())
+			t.execStatement(db, stmt)
+
+		case "query":
+			query := logicQuery{pos: fmt.Sprintf("%s:%d", base, s.line)}
+			// Parse "query error <regexp>"
+			if m := errorRE.FindStringSubmatch(s.Text()); len(m) == 2 {
+				query.expectErr = m[1]
+			} else {
+				// TODO(pmattis): Parse "query <type-string> <sort-mode> <label>". The
+				// type string specifies the number of columns and their types: T for
+				// text, I for integer and R for floating point. The sort mode is one
+				// of "nosort", "rowsort" or "valuesort". The default is "nosort".
+				//
+				// The label is optional. If specified, the test runner stores a hash
+				// of the results of the query under the given label. If the label is
+				// reused, the test runner verifieds that the results are the
+				// same. This can be used to verify that two or more queries in the
+				// same test script that are logically equivalent always generate the
+				// same output.
+			}
+			var buf bytes.Buffer
+			for s.Scan() {
+				t := s.Text()
+				if t == "----" {
+					break
+				}
+				fmt.Fprintln(&buf, s.Text())
```

Comment:
s/s.Text()/t/ (or whatever you rename the binding to to avoid the confusing shadow)